### PR TITLE
feat: support more model types and builtin tools on aws/sagemaker

### PR DESCRIPTION
**Pull request by ybalbert001**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. Close issue syntax: `Fixes #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

Fixes 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A
- [x] Test B

### CODE REVIEW

File: api/core/model_runtime/model_providers/sagemaker/llm/llm.py [hunk 1]
```diff
@@ -25,12 +44,153 @@
 
 logger = logging.getLogger(__name__)
 
+def inference(predictor, messages:list[dict[str,Any]], params:dict[str,Any], stop:list, stream=False):
+    """
+    根据给定的模型名称和端点名称，对消息进行推理。
```
Comment (crazywoola) [inline api/core/model_runtime/model_providers/sagemaker/llm/llm.py]: All comments should be in English

File: api/core/model_runtime/model_providers/sagemaker/llm/llm.py [hunk 2]
```diff
@@ -25,12 +44,153 @@
...
+        """
+            handle normal chat generate response
+        """
+        resp_obj = json.loads(resp.decode('utf-8'))
+        resp_str = resp_obj.get('choices')[0].get('message').get('content')
+
+        if len(resp_str) == 0:
+            raise InvokeServerUnavailableError("Empty response")
+
+        # assistant_message = resp.choices[0].message
```
Comment (crazywoola) [inline api/core/model_runtime/model_providers/sagemaker/llm/llm.py]: Remove commented code here.

File: api/core/model_runtime/model_providers/sagemaker/tts/tts.py
Comment (crazywoola) [inline api/core/model_runtime/model_providers/sagemaker/tts/tts.py:1]: Tested with Yi

```
Node LLM run failed: [sagemaker] Error: 'NoneType' object has no attribute 'predict_stream'
```

image

Comment (crazywoola) [inline api/core/model_runtime/model_providers/sagemaker/tts/tts.py:1]: Also can not save the configurations in Agent mode.

image

image

```
Traceback (most recent call last):
  File "/Users/minibanana/Program/projects/dify/api/.venv/lib/python3.10/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/minibanana/Program/projects/dify/api/.venv/lib/python3.10/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
  File "/Users/minibanana/Program/projects/dify/api/.venv/lib/python3.10/site-packages/flask_restful/__init__.py", line 489, in wrapper
    resp = resource(*args, **kwargs)
  File "/Users/minibanana/Program/projects/dify/api/.venv/lib/python3.10/site-packages/flask/views.py", line 110, in view
    return current_app.ensure_sync(self.dispatch_request)(**kwargs)  # type: ignore[no-any-return]
  File "/Users/minibanana/Program/projects/dify/api/.venv/lib/python3.10/site-packages/flask_restful/__init__.py", line 604, in dispatch_request
    resp = meth(*args, **kwargs)
  File "/Users/minibanana/Program/projects/dify/api/controllers/console/setup.py", line 65, in decorated
    return view(*args, **kwargs)
  File "/Users/minibanana/Program/projects/dify/api/libs/login.py", line 93, in decorated_view
    return current_app.ensure_sync(func)(*args, **kwargs)
  File "/Users/minibanana/Program/projects/dify/api/controllers/console/wraps.py", line 22, in decorated
    return view(*args, **kwargs)
  File "/Users/minibanana/Program/projects/dify/api/controllers/console/app/wraps.py", line 48, in decorated_view
    return view_func(*args, **kwargs)
  File "/Users/minibanana/Program/projects/dify/api/controllers/console/app/model_config.py", line 38, in post
    new_app_model_config = new_app_model_config.from_model_config_dict(model_configuration)
  File "/Users/minibanana/Program/projects/dify/api/models/model.py", line 392, in from_model_config_dict
    self.model = json.dumps(model_config['model']) \
  File "/Users/minibanana/.pyenv/versions/3.10.14/lib/python3.10/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/Users/minibanana/.pyenv/versions/3.10.14/lib/python3.10/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/Users/minibanana/.pyenv/versions/3.10.14/lib/python3.10/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/Users/minibanana/Program/projects/dify/api/.venv/lib/python3.10/site-packages/frozendict/__init__.py", line 31, in default
    return BaseJsonEncoder.default(self, obj)
  File "/Users/minibanana/.pyenv/versions/3.10.14/lib/python3.10/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type LLMMode is not JSON serializable
```

**Review by crazywoola — CHANGES_REQUESTED** [review]
TTS is working for me.

File: api/core/model_runtime/model_providers/sagemaker/tts/tts.py
Comment (ybalbert001) [reply api/core/model_runtime/model_providers/sagemaker/tts/tts.py:1]: could not reproduce it

File: api/core/model_runtime/model_providers/sagemaker/llm/llm.py (hunk 1 above)
Comment (ybalbert001) [reply api/core/model_runtime/model_providers/sagemaker/llm/llm.py]: fixed it

File: api/core/model_runtime/model_providers/sagemaker/llm/llm.py (hunk 2 above)
Comment (ybalbert001) [reply api/core/model_runtime/model_providers/sagemaker/llm/llm.py]: fixed it

File: api/core/model_runtime/model_providers/sagemaker/tts/tts.py
Comment (ybalbert001) [reply api/core/model_runtime/model_providers/sagemaker/tts/tts.py:1]: fixed it

**Review by crazywoola — APPROVED** [review]

**Pull request merged**